vllm/model_executor/models/paligemma.py

```diff
@@ -46,95 +49,161 @@ class PaliGemmaImageEmbeddingInputs(TypedDict):
-        super().__init__()
 
-        self.linear = nn.Linear(vision_hidden_size, projection_dim, bias=True)
+class PaliGemmaMultiModalProcessor(
+        BaseMultiModalProcessor[PaliGemmaProcessingInfo]):
 
-    def forward(self, image_features: torch.Tensor) -> torch.Tensor:
-        hidden_states = self.linear(image_features)
-        return hidden_states
+    def _call_hf_processor(
+        self,
+        prompt: str,
+        mm_data: Mapping[str, object],
+        mm_kwargs: Mapping[str, object],
+    ) -> BatchFeature:
+        tokenizer = self.info.get_tokenizer()
+        if not mm_data:
+            prompt_ids = tokenizer.encode(prompt)
+            return BatchFeature(dict(input_ids=[prompt_ids]), tensor_type="pt")
+
+        return super()._call_hf_processor(
+            prompt=prompt,
+            mm_data=mm_data,
+            mm_kwargs=mm_kwargs,
+        )
 
+    def _get_mm_fields_config(
+        self,
+        hf_inputs: BatchFeature,
+        hf_processor_mm_kwargs: Mapping[str, object],
+    ) -> Mapping[str, MultiModalFieldConfig]:
+        return dict(pixel_values=MultiModalFieldConfig.batched("image"))
 
-@MULTIMODAL_REGISTRY.register_image_input_mapper()
-@MULTIMODAL_REGISTRY.register_max_image_tokens(get_max_paligemma_image_tokens)
-@INPUT_REGISTRY.register_dummy_data(dummy_data_for_paligemma)
-@INPUT_REGISTRY.register_input_processor(input_processor_for_paligemma)
+    def _get_prompt_updates(
+        self,
+        mm_items: MultiModalDataItems,
+        hf_processor_mm_kwargs: Mapping[str, object],
+        out_mm_kwargs: MultiModalKwargs,
+    ) -> list[PromptReplacement]:
+        hf_config = self.info.get_hf_config()
+        image_token_id = hf_config.image_token_index
+
+        tokenizer = self.info.get_tokenizer()
+        num_image_tokens = self.info.get_num_image_tokens()
+        image_tokens = [image_token_id] * num_image_tokens
+
+        bos_token_id = tokenizer.bos_token_id
+        assert isinstance(bos_token_id, int)
+
+        # Paligemma 1 and 2 have different tokenizer.add_bos_token
+        # Replace <bos> with <bos> + <image>*n + <bos> for Paligemma 1
+        # Insert <image>*n + <bos> for Paligemma 2
+        if tokenizer.add_bos_token:
+            return [
+                PromptReplacement(
+                    modality="image",
+                    target=[bos_token_id],
+                    replacement=PromptUpdateDetails(
+                        full=[bos_token_id] + image_tokens + [bos_token_id],
+                        features=image_tokens,
+                    ),
+                )
+            ]
+        else:
+            return [
+                PromptInsertion(
+                    modality="image",
+                    target=[],
+                    insertion=PromptUpdateDetails(
+                        full=image_tokens + [bos_token_id],
+                        features=image_tokens,
+                    ),
+                )
+            ]
```

Comment:
```suggestion
        # Paligemma 1 and 2 have different tokenizer.add_bos_token
        # Replace <bos> with <bos> + <image>*n + <bos> for Paligemma 1
        # Insert <image>*n + <bos> for Paligemma 2
        return [
            PromptReplacement(
                modality="image",
                target=PromptIndexTargets.prefix([bos_token_id] if tokenizer.add_bos_token else []),
                replacement=PromptUpdateDetails(
                    full=image_tokens + [bos_token_id],
                    features=image_tokens,
                ),
            )
        ]
```

Simplify things a bit